bdk_electrum_example/src/electrum.rs

```diff
@@ -0,0 +1,150 @@
+    BlockId, TxHeight,
+};
+use electrum_client::{Client, Config, ElectrumApi};
+
+pub struct ElectrumClient {
+    client: Client,
+}
+
+impl ElectrumClient {
+    pub fn new(url: &str) -> Result<Self> {
+        let client = Client::from_config(url, Config::default())?;
+        Ok(Self { client })
+    }
+}
+
+impl Deref for ElectrumClient {
+    type Target = Client;
+    fn deref(&self) -> &Self::Target {
+        &self.client
+    }
+}
+
+impl Broadcast for ElectrumClient {
+    type Error = electrum_client::Error;
+    fn broadcast(&self, tx: &bdk_core::bitcoin::Transaction) -> Result<(), Self::Error> {
+        let _ = self.client.transaction_broadcast(tx)?;
+        Ok(())
+    }
+}
+
+/// We will detect reorg if it has depth less than this
+const REORG_DETECTION_DEPTH: u32 = 100;
+const DEFAULT_STOP_GAP: usize = 10;
+
+impl ElectrumClient {
+    /// Fetch latest block height.
+    pub fn get_height(&self) -> Result<u32> {
+        // TODO: unsubscribe when added to the client, or is there a better call to use here?
+        Ok(self
+            .client
+            .block_headers_subscribe()
+            .map(|data| data.height as u32)?)
+    }
+
+    /// Scan for a list of scripts, and create an initial [`ChainGraph`] candidate update.
+    /// This update will only contain [`Txid`]s in SparseChain, and no actual transaction data.
+    ///
+    /// User needs to fetch the required transaction data and update the [`ChainGraph`] before applying it.
+    pub fn spk_txid_scan(
+        &self,
+        spks: impl Iterator<Item = Script>,
+        last_known_height: Option<u32>,
+    ) -> Result<SparseChain> {
+        let mut dummy_keychains = BTreeMap::new();
+        dummy_keychains.insert((), spks.enumerate().map(|(i, spk)| (i as u32, spk)));
+
+        Ok(self
+            .wallet_txid_scan(dummy_keychains, None, last_known_height)?
+            .0)
+    }
+
+    /// Scan for a keychain tracker, and create an initial [`KeychainScan`] candidate update.
+    /// This update will only contain [`Txid`]s in SparseChain, and no actual transaction data.
+    ///
+    /// User needs to fetch the required transaction data and update the [`KeychainScan`] before applying it.
+    pub fn wallet_txid_scan<K: Ord + Clone>(
+        &self,
+        scripts: BTreeMap<K, impl Iterator<Item = (u32, Script)>>,
+        stop_gap: Option<usize>,
+        last_known_height: Option<u32>,
+    ) -> Result<(SparseChain, BTreeMap<K, u32>)> {
+        let mut sparse_chain = SparseChain::default();
+
+        // 1. Create checkpoint data from last_known_height - REORG_DEPTH
+        // Reorg situation will he handled at the time of applying this KeychainScan.
+        // If there's reorg deeper than the assumed depth, update process will throw error.
+        let current_height = self.get_height()?;
+        let check_from = last_known_height
+            .map(|ht| ht.saturating_sub(REORG_DETECTION_DEPTH))
+            .unwrap_or(0);
```

Comment:
If the tip at `last_known_height` has changed then this will always lead to an error in the case the checkpoint before `last_known_height` is more than 100 blocks away. Why not do it the way we did in esplora?

Reply:
Yes, I was going with the assumption that `SparseChain` will always maintain consecutive sequence of blocks.. 

The rationale behind that was, in sync code we are always adding sequential blocks in checkpoints, and pruning is happening from the start of the chain.. So by that logic we should never have the last common at 100+ blocks away from `last_known_height`? As long as that's true, the other way of throwing error here is more than 100 blocks of reorg, which is unlikely..

But if my assumption is wrong then yes this should not be done in this way, and the esplora way is better. This one is clearly less elegant than the esplora way, but went in this route to reduce passing in any refs from the tracker.. The chain was the only refs required to pass in in the sync function..

Reply:
> Yes, I was going with the assumption that `SparseChain` will always maintain consecutive sequence of blocks..

This is definitely not the case. Perhaps it's confusing to call it a "chain". It's meant to be a view of a chain that is sparse not a chain that is sparse.

> But if my assumption is wrong then yes this should not be done in this way, and the esplora way is better. This one is clearly less elegant than the esplora way, but went in this route to reduce passing in any refs from the tracker.. The chain was the only refs required to pass in in the sync function..

You can just clone the checkpoints to avoid taking a ref. There might be other ways to design this process but the way it works is to limit the number of checkpoints in the sparse chain with `checkpoint_limit` so they can be quickly cloned and passed in here without taking a ref. This seems to be ideal since the your checkpoint limit will be a function of your confidence of the persistence of the underlying blockchain and the memory available on your platform -- these are not things that can change at runtime and should be fixed in the config.

Reply:
Okay.. Ya that makes sense to me.. So I will revert it back to the esplora way..